librepo/downloader.c

```diff
@@ -691,64 +691,74 @@ select_suitable_mirror(LrDownload *dd,
-            && target->handle->offline
-            && c_mirror->mirror->protocol != LR_PROTOCOL_FILE)
-        {
-            // Skip each url that doesn't have "file://" or "file:" prefix
-            g_debug("%s: Skipping mirror %s - Offline mode enabled",
-                    __func__, mirrorurl);
-            continue;
-        }
+            if (target->handle
+                && target->handle->offline
+                && c_mirror->mirror->protocol != LR_PROTOCOL_FILE)
+            {
+                if (!maximal_errors) {
+                // Skip each url that doesn't have "file://" or "file:" prefix
+                g_debug("%s: Skipping mirror %s - Offline mode enabled", __func__, mirrorurl);
+                }
+                continue;
+            }
 
-        if (c_mirror->successful_transfers == 0 &&
-            dd->allowed_mirror_failures > 0 &&
-            c_mirror->failed_transfers >= dd->allowed_mirror_failures)
-        {
-            // Skip bad mirrors
-            g_debug("%s: Skipping bad mirror (%d failures and no success): %s",
-                    __func__, c_mirror->failed_transfers, mirrorurl);
-            continue;
-        }
+            at_least_one_suitable_mirror_found = TRUE;
 
-        at_least_one_suitable_mirror_found = TRUE;
+            // Number of transfers which are downloading from the mirror
+            // should always be lower or equal than maximum allowed number
+            // of connection to a single host.
+            assert(dd->max_connection_per_host == -1 ||
+                c_mirror->running_transfers <= dd->max_connection_per_host);
 
-        // Number of transfers which are downloading from the mirror
-        // should always be lower or equal than maximum allowed number
-        // of connection to a single host.
-        assert(dd->max_connection_per_host == -1 ||
-               c_mirror->running_transfers <= dd->max_connection_per_host);
+            // Init max of allowed parallel connections from config
+            init_once_allowed_parallel_connections(c_mirror, dd->max_connection_per_host);
 
-        // Init max of allowed parallel connections from config
-        init_once_allowed_parallel_connections(c_mirror, dd->max_connection_per_host);
+            // Check number of connections to the mirror
+            if (is_parallel_connections_limited_and_reached(c_mirror))
+            {
+                continue;
+            }
 
-        // Check number of connections to the mirror
-        if (is_parallel_connections_limited_and_reached(c_mirror))
-        {
-            continue;
+            // This mirror looks suitable - use it
+            *selected_mirror = c_mirror;
+            return TRUE;
         }
-
-        // This mirror looks suitable - use it
-        *selected_mirror = c_mirror;
-        return TRUE;
-    }
+    } while (g_slist_length(target->tried_mirrors) < dd->allowed_mirror_failures &&
+    ++maximal_errors < dd->allowed_mirror_failures);
```

Comment:
Might be personal preference, but why not use a `while` instead of `do ... while`, makes it easier to read and there's no benefit to `do` here I think.

Also, the whole mirror-picking mechanism seems to be complex already and adding one more loop around it pushes it to being quite convoluted.

Why not make it simply loop over the available mirrors until either success or number of retries is reached?

```C++
GSList *elem = target->lrmirrors;
int num_retires = 0;
while (elem != NULL && num_retires <= dd->allowed_mirror_failures) {
    if (try_mirror(...)) {
        break;
    }

    num_retries++;
    elem = g_slist_next(elem);
    // loop around in case we reach the end of the mirror list
    if (elem == NULL) {
        elem = target->lrmirrors;
    }
}
```

Reply:
I am sorry but I have a feeling that the proposed approach will not deliver the required behaviour.

Reply:
What _is_ the required behaviour though? I have a feeling the current code is unnecessarily complex.

Reply:
@lukash From my point of view do {} while () loop is more readable that proposed change. The behaviour should allow to use all mirrors first. Then to allow of reuse of limited number of mirrors.
The construct directly says it. Please could you specify whatever it is no go request?

Reply:
> @lukash From my point of view do {} while () loop is more readable that proposed change.

Readability is subjective, but I don't think you can really say that your code is more readable than what I proposed. You have two nested loops there with several more nested `if`s checking multiple conditions. Granted my example is simplified and will surely need more stuff added, it is a simple logical flow of what you'd expect the function to do. So I find it hard to believe you think your code is more readable...

> The behaviour should allow to use all mirrors first. Then to allow of reuse of limited number of mirrors.

This is too vague and in fact my example can do exactly the same (by adding a single `if`). I believe that if you want to merge code like this, you should be able to describe with words what it is supposed to be doing and actually put that into a comment. Specifically focusing on the reasons why it actually needs to be complex. One thing is making it easier to understand for others, the other thing is, by adding complexity, you are adding more corner cases and more possibilities for bugs. (And we should really be avoiding this, that's why I'm having this argument)

Actually, you have two nested loops there, while there is no code in the outer loop (apart from incrementing `maximal_errors`) and both loops iterate over the same list of items. So I think it can be proven theoretically you can achive the same with a single loop :slightly_smiling_face:.

> The construct directly says it. Please could you specify whatever it is no go request?

I'm not sure I understand what you mean here, but if you're asking whether I insist on changing this in the PR, I think so, until I understand the reasons it needs to be like this...

Cheers!

Reply:
I added comment to provide additional hint why the loop is there. I don't have a good feeling to change the loop system here. Especially in case that the change request is marked as "Might be personal preference".
But I have a different question: Shall we add some wait step after the first iteration?

Reply:
It is not just a "Might be personal preference", so please don't dismiss it as that.

I've fixed up your comment and added a couple more while reviewing the code, please have a look and if you don't mind, squash it with your commit: https://github.com/lukash/librepo/commit/8d4dd9f8618fa64b37d17c0e65e9180294fa9d71

As for the wait, if you mean to add it to this loop you've added to select_suitable_mirror(), then I'd say certainly not, as that function only selects the mirror for the next download, so IMHO you'd just be messing up the flow by adding random waits. But I'm still barely understanding the code, so...